### PR TITLE
[WIP] HttpUrl - Punycode and Un-encoded params

### PR DESCRIPTION
**Pull request by yschimke**
Exploring how this would work and what edge cases there are.  As I accept user input in my own project, these end up being displayed in a very non-intuitive way.

Chrome will happily show 

https://www.google.com/search?q=рф
or 
http://ουτοπία.δπθ.gr

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 1]
```diff
@@ -972,6 +975,44 @@ static HttpUrl getChecked(String url) throws MalformedURLException, UnknownHostE
     return PublicSuffixDatabase.get().getEffectiveTldPlusOne(host);
   }
 
+  public static String displayHost(Builder.DisplayMode displayMode, String host) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: I don’t think DisplayMode is right as public API. It seems like an implementation detail to me.

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 2]
```diff
@@ -972,6 +975,44 @@ static HttpUrl getChecked(String url) throws MalformedURLException, UnknownHostE
...
+        return isDisplaySafe(unicodeHost) ? unicodeHost : host;
+      case UNICODE:
+        return IDN.toUnicode(host, 0);
+      default:
+        return host;
+    }
+  }
+
+  /**
+   * Returns the unicode host if the host is considered IDN safe.
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Provide an example?

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: ALSO – I believe in Chrome the display URLs take a Locale, and only use Unicode when the locales match. Should this do likewise?

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 3]
```diff
@@ -1263,6 +1304,10 @@ public HttpUrl build() {
     }
 
     @Override public String toString() {
+      return toString(DisplayMode.RAW);
+    }
+
+    private String toString(DisplayMode displayMode) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: nit: package-private? As-is this generates a trampoline method

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 4]
```diff
@@ -1295,7 +1340,12 @@ public HttpUrl build() {
 
       if (encodedQueryNamesAndValues != null) {
         result.append('?');
-        namesAndValuesToQueryString(result, encodedQueryNamesAndValues);
+        if (displayMode != DisplayMode.RAW) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: HMMM . . . these URLs are ambiguous aren’t they?

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Do browsers do this?

File: okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java [hunk 5]
```diff
@@ -1507,4 +1511,97 @@
     assertNull(HttpUrl.parse("https://localhost").topPrivateDomain());
     assertNull(HttpUrl.parse("https://127.0.0.1").topPrivateDomain());
   }
+
+  @Test public void testDisplayStringPunyCodeNonIdentifier() {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java]: nice tests

File: okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java [hunk 6]
```diff
@@ -1507,4 +1511,97 @@
...
+    assertReversible("http://ακρόπολητώρα.org/file.html", "ακρόπολητώρα.org");
+
+    assertReversible("http://☃.net/file.html", "");
+    // mixed characters in wikipedia
+    assertReversible("http://wіkіреdіа.org/file.html", "");
+
+    assertReversible("http://правительство.рф/search?q=рф", "правительство.рф");
+    assertReversible("http://en.wikipedia.org/search?q", "q");
+    assertReversible("http://en.wikipedia.org/search?q=a b", "q=a b");
+    // TODO how to handle?
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java]: What do browsers do?

File: okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java [hunk 7]
```diff
@@ -1507,4 +1511,97 @@
...
+    HttpUrl originalUrl = HttpUrl.parse(originalUrlString);
+
+    String urlToString = originalUrl.toString();
+    String urlToDisplayString = originalUrl.toDisplayString();
+
+    assertTrue("Couldn't find pattern '" + toStringCheck + "' in '" + urlToDisplayString + "'",
+        Pattern.compile(toStringCheck).matcher(urlToDisplayString).find());
+
+    HttpUrl reparsedUrl = HttpUrl.parse(urlToString);
+    HttpUrl reparsedDisplayUrl = HttpUrl.parse(urlToDisplayString);
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java]: I really like this approach

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 2 above)
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: I'll investigate.

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 4 above)
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Yes, somewhat.  They seem to do some rewriting, but still maintain the reversibility.  I know this is tricky code, so I'm not in a rush to land this, but I think it's useful functionality.

**Comment by yschimke** [comment]
Thanks for the review, my main question was whether this fits within OkHttp as I think it should.  It seems like if I can get it right you aren't against this.

**Comment by swankjesse** [comment]
Yes, I definitely want this. Biggest challenge is trying to make it both usable and safe. Following the leads of Chrome, Firefox and Safari is the way to get there.

**Comment by swankjesse** [comment]
By the way, I really like reversibility as a goal here. If it’s reversible we’re definitely on the right track.

**Comment by yschimke** [comment]
Will reopen when I get back to this - been busy recently.

**Pull request closed without merging**